File: lol_gen_alg_python/genetic_algorithm/team_comp.py
import random
import logging
import math
import copy
from genetic_algorithm.champion import Champion
logger = logging.getLogger()


class TeamComp:
    def __init__(self, champion_information: dict, champions: [Champion],
                 mutation_rate=0.9):
        """
        Create a new team comp. Champions either supplied or list of avilable
        champions is provided
        :param champion_information: information for available champions
        :param champions: Given champion list for the team Comp
        :param mutation_rate: rate to mutate. 9% by default
        """
        if champions:
            self.champions = champions
        else:
            self.champions = TeamComp.generate_team_comp(champion_information)
        self.fitness = -1
        self.mutation_rate = mutation_rate

    def mutate(self, available_champions: dict, individual=-1):
        """
        We mutate one of the champions. This means replacing a current champion.
        The list of available champions is all champions minus base team comp
        and this current team comp If our mutation rate is .9,
        and the random is less or equal then we mutate, thus simulating a
        90% mutation rate.
        :return:
        """
        if random.uniform(0, 1) <= self.mutation_rate:

            # Then we mutate one of the champions!
            index = random.randint(0, 4)
            current_champion_ids = [x.id for x in self.champions]
            new_champion = TeamComp.make_new_champion(
                available_champions, current_champion_ids)

            logger.debug(f"Replacing individual {individual}'s "
                         f"{self.champions[index].name} with {new_champion.name}")
            self.champions[index] = new_champion

    def calculate_fitness(self, base_team_composition: [Champion]):
        avg_winrate = 0
        total_calculated = 0

        for your_champion in self.champions:
            for enemy_champion in base_team_composition:
                win_rate = your_champion.check_winrate(enemy_champion.id)
                if win_rate:
                    avg_winrate += win_rate
                    total_calculated += 1

        avg_winrate = avg_winrate/total_calculated
        # Log base 10 calculation supports teams with more available winrates
        fitness = avg_winrate * math.log(total_calculated, 10)
        self.fitness = fitness

    def is_on_team(self, champion_id) -> bool:
        """
        Given a champion_id, check if it is on current team
        :param champion_id: ID to check for
        :return: true if on team, else false
        """
        for champion in self.champions:
            if champion.id == champion_id:
                return True
        return False

    @staticmethod
    def generate_team_comp(champion_details: dict) -> [Champion]:
        """
        Given all matchup information. Create a random team comp
        :param champion_details: Dict containing all matchup information for all champions
        :return: list of Champions
        """
        champions = []
        ids_chosen = []
        for i in range(5):
            new_champion = TeamComp.make_new_champion(champion_details, ids_chosen)
            champions.append(new_champion)
            ids_chosen.append(new_champion.id)
        return champions

    @staticmethod
    def make_new_champion(champion_details: dict, current_ids: list) -> Champion:
        """
        Make a new RANDOM champion
        :param current_ids: Current IDs
        :param champion_details: Dict containing all mactchup information for all champions
        :return: A new champion
        """
        # Make sure champion selected isn't in the list of current champions
        champion_id = random.choice(list(champion_details))
        while champion_id in current_ids and len(current_ids) > 0:
            champion_id = random.choice(list(champion_details))
        champion_details = champion_details.get(champion_id)
        return Champion(
            champion_id, champion_details.get("name"),
            champion_details.get("win_rates"))

File: lol_gen_alg_frontend/src/gen_alg_logic/team_comp.js
import Champion from './champion.js';

export default class TeamComp {

	constructor(champ_info, champions=null, mutation_rate=0.9){

		if (champions) {
			this.champions = champions;
		}
		else{
			this.champions = TeamComp.generate_team_comp(champ_info);
		}
		this.fitness = -1;
		this.mutation_rate = mutation_rate;
	}

	mutate(available_champs){
		if (Math.random() <= this.mutation_rate) {
			let index = getRandomInt(5);
			let current_ids = [];
			this.champions.forEach(function(champ, i){
				current_ids.push(champ.id);
			})
			let new_champion = TeamComp.make_new_champion(
				available_champs, current_ids)
			this.champions[index] = new_champion;
		}
	}

	calculate_fitness(enemy_team_comp){
		/*
    For all champions on your team, check their winrates
    against the given enemy team's champions. Take average of all
    available win rates (some might not have information)
    multiply that by the log base 10 of all with information.
    This will make teams with more information have a better fitness.
		*/
		var avg_winrate = 0;
		var total_calculated = 0;
		this.champions.forEach(function(your_champion, your_index){
			enemy_team_comp.forEach(function(enemy_champion,enemy_index){
				let win_rate = your_champion.check_winrate(enemy_champion.id);
				if (win_rate) {
					avg_winrate += win_rate;
					total_calculated ++;
				}
			})
		})
		let fitness = avg_winrate/total_calculated * Math.log10(total_calculated);
		this.fitness = fitness;
	}

	is_on_team(champion_id){
		/*
    Given a champion_id, check if it is on current team
    :param champion_id: ID to check for
    :return: true if on team, else false
		*/
		this.champions.forEach(function(item,index){
			if (champion_id === item.id){
				return true;
			}
		})
		return false;
	}

	static generate_team_comp(champ_info){
		/*
    Given all matchup information. Create a random team comp
    :param champ_info: Dict containing all matchup information for all champions
    :return: list of Champions
		*/
		let champions = [];
		let ids_chosen = [];
		for (let i = 0; i < 5; i++) {
			let new_champion = TeamComp.make_new_champion(champ_info, ids_chosen);
			champions.push(new_champion);
			ids_chosen.push(new_champion.id);
		}
		return champions;
	}

	static make_new_champion(champ_info, current_ids){
		/*
    Make a new RANDOM champion
    :param current_ids: Current IDs
    :param champ_info: Dict containing all mactchup information for all champions
    :return: A new champion
		*/
		var champion_id = get_random_key(champ_info);
				// Get a unique champion not in current team_comp
        while (current_ids.includes(champion_id) && current_ids.length > 0) {
            champion_id = get_random_key(champ_info);
        }
        let champion_details = champ_info[champion_id];
        return new Champion(
        	champion_id, champion_details.name, 
        	champion_details.win_rates,
        	champion_details.overall_win_rate,
      		champion_details.total_games
        )

	}
}

function get_random_key(dict){
    var keys = Object.keys(dict);
		return keys[ keys.length * Math.random() << 0];
}

function getRandomInt(max) {
 	return Math.floor(Math.random() * Math.floor(max));
}
